src/deadline/job_attachments/asset_sync.py

```diff
@@ -105,6 +118,333 @@ def __init__(
 
         self.hash_alg: HashAlgorithm = self.manifest_model.AssetManifest.get_default_hash_alg()
 
+    @staticmethod
+    def generate_dynamic_path_mapping(
+        session_dir: Path,
+        attachments: Attachments,
+    ) -> dict[str, PathMappingRule]:
+        """
+        Compute path mapping rules that are relative to the given session directory.
+
+        Args:
+            session_dir: path to the current session directory
+            attachments: an object that holds all input assets for the job.
+
+        Returns: a dictionary of local roots for each asset root, used for path mapping.
+        """
+        mapped_path: dict[str, PathMappingRule] = dict()
+
+        for manifest_properties in attachments.manifests:
+            if not manifest_properties.fileSystemLocationName:
+                dir_name: str = _get_unique_dest_dir_name(manifest_properties.rootPath)
+                local_root = str(session_dir.joinpath(dir_name))
+                mapped_path[manifest_properties.rootPath] = PathMappingRule(
+                    source_path_format=manifest_properties.rootPathFormat.value,
+                    source_path=manifest_properties.rootPath,
+                    destination_path=local_root,
+                )
+
+        return mapped_path
+
+    @staticmethod
+    def get_local_destination(
+        manifest_properties: ManifestProperties,
+        dynamic_mapping_rules: dict[str, PathMappingRule] = {},
+        storage_profiles_path_mapping_rules: dict[str, str] = {},
+    ) -> str:
+        """
+        Args:
+            manifest_properties: manifest properties to search local destination for.
+            dynamic_mapping_rules: manifest root path to worker host destination mapping relative to local session.
+            storage_profiles_path_mapping_rules: a dict of source path -> destination path mappings.
+
+        Returns: local destination corresponding to the given manifest properties.
+        Raises: AssetSyncError If no path mapping rule is found for the given root path.
+        """
+        root_path = manifest_properties.rootPath
+
+        if manifest_properties.fileSystemLocationName:
+            local_destination = storage_profiles_path_mapping_rules.get(root_path)
+        else:
+            path_mapping: Optional[PathMappingRule] = dynamic_mapping_rules.get(root_path)
+            local_destination = path_mapping.destination_path if path_mapping else None
+
+        if local_destination:
+            return local_destination
+        else:
+            raise AssetSyncError(
+                "Error occurred while attempting to sync input files: "
+                f"No path mapping rule found for the source path {manifest_properties.rootPath}"
+            )
+
+    def aggregate_asset_root_manifests(
```

Comment:
Question, is there any member variable to access here? can this also be static?

Reply:
Can be done as a follow up since we are still working on this.

Reply:
Yup there are few member variables this function is using such as session and farm id. This file could use more improvements, will keep an eye out for additional refactoring opportunities as I work on it.